libraries/pet_service/pet_service.lua

```diff
@@ -0,0 +1,176 @@
+local math = require('math')
+local ffi = require('ffi')
+
+local packet = require('packet')
+local server = require('shared.server')
+local struct = require('struct')
+
+local item_type = struct.struct({
+    item_id = {struct.uint16},
+    raw     = {struct.uint8},
+})
+
+local data = server.new(struct.struct({
+    index           = {struct.uint16},
+    id              = {struct.uint32},
+    name            = {struct.string(0x10)},
+    owner_index     = {struct.uint16},
+    owner_id        = {struct.uint32},
+    target_id       = {struct.uint32},
+    hp_percent      = {struct.uint8},
+    mp_percent      = {struct.uint8},
+    tp              = {struct.uint32},
+    active          = {struct.bool},
+    automaton       = {struct.struct({
+        active              = {struct.bool},
+        head                = {item_type},
+        frame               = {item_type},
+        attachments         = {item_type[12]},
+        available_heads     = {struct.bitfield(4)},
+        available_frames    = {struct.bitfield(4)},
+        available_attach    = {struct.bitfield(32)},
+        name                = {struct.string(0x10)},
+        hp                  = {struct.uint16},
+        hp_max              = {struct.uint16},
+        mp                  = {struct.uint16},
+        mp_max              = {struct.uint16},
+        melee               = {struct.uint16},
+        melee_max           = {struct.uint16},
+        ranged              = {struct.uint16},
+        ranged_max          = {struct.uint16},
+        magic               = {struct.uint16},
+        magic_max           = {struct.uint16},
+        str                 = {struct.uint16},
+        str_modifier        = {struct.uint16},
+        dex                 = {struct.uint16},
+        dex_modifier        = {struct.uint16},
+        vit                 = {struct.uint16},
+        vit_modifier        = {struct.uint16},
+        agi                 = {struct.uint16},
+        agi_modifier        = {struct.uint16},
+        int                 = {struct.uint16},
+        int_modifier        = {struct.uint16},
+        mnd                 = {struct.uint16},
+        mnd_modifier        = {struct.uint16},
+        chr                 = {struct.uint16},
+        chr_modifier        = {struct.uint16},
+    })}
+}))
+
+packet.incoming:register_init({
+    [{0x037}] = function(p) -- While this packet is mostly player data, it does occassionally update the pet index when no other pet related packet is sent. For example, when moving into zones where the pet is supressed, such as cities and towns, this packet will set the pet index to 0.
+        data.index = p.pet_index
+        if p.pet_index and p.pet_index ~= 0 then
```

Comment:
`p.pet_index` should never be `nil`, so the first check could be removed (elsewhere as well).